# Bayesian Network for Inference

See also:

https://www.coursera.org/specializations/probabilistic-graphical-models

Objective: Implement a simple Bayesian network for inference.

Description:
Write a Python script to:
1. Define a Bayesian network structure for a given problem (e.g., disease diagnosis).
2. Specify the conditional probability tables (CPTs) for each node.
3. Perform inference to calculate the probability of an outcome given some evidence.
4. Output the inferred probabilities.

Requirements:
- Use the `pgmpy` library for Bayesian networks.
- Ensure the Bayesian network is correctly defined and the inference results are accurate.


In a Bayesian Network, each node represents a random variable, and the network structure encodes the conditional dependencies between these variables. The edges in the network indicate direct dependencies, and the strength of these dependencies is quantified by Conditional Probability Distributions (CPDs).

### Bayesian Networks and Bayes' Rule

**Bayes' Rule:**

Bayes' rule is a fundamental theorem in probability theory and statistics that describes how to update the probability of a hypothesis based on new evidence:

$$
P(H | E) = \frac{P(E | H) \cdot P(H)}{P(E)}
$$

Where:
- \( P(H | E) \) is the posterior probability of the hypothesis \( H \) given the evidence \( E \).
- \( P(E | H) \) is the likelihood of observing the evidence \( E \) given that \( H \) is true.
- \( P(H) \) is the prior probability of the hypothesis \( H \).
- \( P(E) \) is the marginal probability of the evidence \( E \).

### Nodes in a Bayesian Network

In a Bayesian Network:
- Each node represents a random variable.
- Each node \( X \) has a CPD that defines \( P(X | \text{Parents}(X)) \), where \text{Parents}(X) are the parent nodes of \( X \) in the network.
- The CPD specifies how the probability distribution of \( X \) changes given different values of its parent nodes.

### Conditional Dependencies

The structure of the network (i.e., the directed edges) represents the conditional dependencies between variables:
- If there is a directed edge from node \( A \) to node \( B \), then \( B \) is conditionally dependent on \( A \), and the CPD for \( B \) will include \( P(B | A) \).
- The absence of an edge indicates conditional independence between variables given their parents in the network.

### Inference in Bayesian Networks

When performing inference in a Bayesian Network (e.g., computing the probability of a variable given some evidence), algorithms like Variable Elimination or Belief Propagation are used. These algorithms apply Bayes' rule and the chain rule of probability to update and compute the desired probabilities.

### Example from the Code

In your example, you have the following conditional dependencies:
- `Fever` depends on `Flu` and `Cold`.
- `Cough` depends on `Cold`.
- `Body Ache` depends on `Flu`.

Each node's CPD defines how it is conditionally dependent on its parents. For instance, the CPD for `Fever` specifies \( P(Fever | Flu, Cold) \), and inference in this network will involve combining these CPDs using Bayes' rule and the chain rule.

### Summary

- The network as a whole, through its structure and CPDs, represents the joint probability distribution over all variables.
- Inference algorithms use Bayes' rule and other probabilistic rules to compute posterior probabilities based on the network's structure and CPDs.

Bayesian Networks provide a powerful framework for representing and reasoning about uncertain knowledge, capturing complex dependencies among variables through a combination of network structure and probabilistic rules.

# When to use Bayesian Networks

Bayesian Networks are powerful tools for modeling probabilistic relationships and making inferences under uncertainty. They are particularly useful in a variety of scenarios where their unique advantages are most beneficial. Here are some situations in which Bayesian Networks are especially well-suited:

### 1. Modeling Complex Systems with Uncertainty

**Use Case**: Medical Diagnosis, Fault Diagnosis in Systems

- **Why**: Bayesian Networks can effectively model the probabilistic relationships between different variables, making them ideal for domains where uncertainty is inherent. For instance, in medical diagnosis, symptoms and diseases are often probabilistically related.

### 2. Handling Missing Data

**Use Case**: Customer Data Analysis

- **Why**: Bayesian Networks can infer the missing values using the relationships between the observed data. This makes them suitable for datasets where missing data is a common issue.

### 3. Causal Inference

**Use Case**: Policy Making, Root Cause Analysis

- **Why**: Bayesian Networks allow for the representation and reasoning about causal relationships. This can be crucial in understanding the impact of different factors and making decisions based on these causal relationships.

### 4. Sequential Decision Making

**Use Case**: Robotics, Automated Planning

- **Why**: Dynamic Bayesian Networks (an extension of Bayesian Networks) are used for modeling time-dependent processes, making them suitable for applications that involve sequential decision-making under uncertainty.

### 5. Integration of Expert Knowledge

**Use Case**: Expert Systems

- **Why**: Bayesian Networks can integrate expert knowledge with data. This is valuable in fields where expert knowledge is critical but data may be sparse or incomplete.

### 6. Data Fusion

**Use Case**: Sensor Networks, Environmental Monitoring

- **Why**: Bayesian Networks can combine information from multiple sources to provide a coherent probabilistic model, making them useful for data fusion tasks.

### 7. Probabilistic Reasoning

**Use Case**: Natural Language Processing, Recommendation Systems

- **Why**: Bayesian Networks support probabilistic reasoning, which is essential for applications that require making predictions or recommendations based on uncertain information.

### 8. Predictive Modeling

**Use Case**: Financial Risk Assessment, Market Basket Analysis

- **Why**: Bayesian Networks can be used to predict the probability of certain outcomes based on the observed data, making them useful for predictive modeling tasks.

### 9. Troubleshooting and Diagnostics

**Use Case**: IT Systems Management, Manufacturing

- **Why**: Bayesian Networks can diagnose issues by calculating the probabilities of potential causes given observed symptoms, making them ideal for troubleshooting and diagnostics.

### Summary of When to Use Bayesian Networks

- **Uncertainty and Probabilistic Relationships**: Ideal for modeling and reasoning with uncertain and probabilistic relationships between variables.
- **Missing Data**: Effective in handling and inferring missing data.
- **Causal Inference**: Useful for understanding and reasoning about causal relationships.
- **Sequential Decision Making**: Suitable for time-dependent processes and sequential decision-making.
- **Expert Knowledge Integration**: Can integrate expert knowledge with data.
- **Data Fusion**: Combines information from multiple sources.
- **Probabilistic Reasoning**: Supports making predictions and recommendations under uncertainty.
- **Predictive Modeling**: Useful for predicting the probability of certain outcomes.
- **Troubleshooting and Diagnostics**: Ideal for diagnosing issues based on observed symptoms.

### Example Applications

- **Healthcare**: Diagnosing diseases based on symptoms and patient history.
- **Finance**: Assessing the risk of loan defaults.
- **Robotics**: Making decisions in uncertain environments.
- **Environmental Science**: Predicting weather patterns or pollution levels.
- **IT Systems**: Diagnosing faults in complex IT systems.

By leveraging the strengths of Bayesian Networks in these scenarios, you can effectively model complex systems, make informed decisions, and handle uncertainty in a structured and probabilistic manner.

# Bayesian Networks: An Algorithmic Description

## High-Level Intuitions

1. **Probabilistic Relationships**: Bayesian Networks represent the probabilistic relationships between variables. They help model the uncertainty and dependencies in complex systems.
2. **Directed Acyclic Graph (DAG)**: The network structure is a directed acyclic graph where nodes represent random variables and edges represent conditional dependencies.
3. **Inference**: Bayesian Networks allow for efficient probabilistic inference, enabling the calculation of the likelihood of certain outcomes given observed evidence.

## Key Components

1. **Nodes**: Represent random variables (e.g., Flu, Cold, Fever).
2. **Edges**: Represent conditional dependencies between variables (e.g., Flu -> Fever).
3. **Conditional Probability Tables (CPTs)**: Define the probability distribution of a node given its parent nodes.

## Steps to Construct and Use a Bayesian Network

### Step 1: Define the Structure

1. Identify the variables in your problem domain.
2. Determine the dependencies between variables.
3. Create a directed acyclic graph (DAG) where each node represents a variable, and edges represent dependencies.

### Step 2: Specify Conditional Probability Tables (CPTs)

1. For each node, define the CPT, which specifies the probability of each possible state of the node given each possible combination of states of its parent nodes.
2. If a node has no parents, its CPT is simply the prior probability distribution of the node.

### Step 3: Validate the Model

1. Ensure that the structure and CPTs of the Bayesian Network are correctly defined.
2. Check for consistency and completeness in the CPTs.

### Step 4: Perform Inference

1. Choose an inference algorithm (e.g., Variable Elimination, Belief Propagation).
2. Use the algorithm to compute the posterior probabilities of certain variables given observed evidence.

### Step 5: Analyze and Interpret Results

1. Examine the inferred probabilities to draw conclusions about the system being modeled.
2. Use the results to make decisions or further predictions.

## Detailed Algorithm

### Structure Definition

1. **Input**: Set of variables \( V \), set of edges \( E \) representing dependencies.
2. **Output**: Directed acyclic graph \( G \).

Algorithm:
1. Initialize an empty graph \( G \).
2. For each variable \( v_i \in V \):
   - Add a node \( v_i \) to \( G \).
3. For each dependency \( (v_i, v_j) \in E \):
   - Add a directed edge from \( v_i \) to \( v_j \) in \( G \).

### CPT Specification

1. **Input**: Graph \( G \) with nodes and edges, domain knowledge or data for probabilities.
2. **Output**: Conditional probability tables (CPTs) for each node.

Algorithm:
1. For each node \( v_i \) in \( G \):
   - Identify the parent nodes \( \text{Parents}(v_i) \).
   - Define the CPT for \( v_i \), specifying \( P(v_i | \text{Parents}(v_i)) \).

### Model Validation

1. **Input**: Graph \( G \), CPTs for each node.
2. **Output**: Validation result (True/False).

Algorithm:
1. Check that \( G \) is a directed acyclic graph (DAG).
2. For each node \( v_i \):
   - Ensure that the sum of probabilities in the CPT of \( v_i \) equals 1 for each combination of parent states.

### Inference

1. **Input**: Graph \( G \), CPTs, evidence \( E \) (observed values for some nodes).
2. **Output**: Posterior probabilities \( P(Q | E) \) for query nodes \( Q \).

Algorithm:
1. Select an inference algorithm (e.g., Variable Elimination).
2. Initialize factors using the CPTs and evidence.
3. For each query node \( q_i \) in \( Q \):
   - Apply the inference algorithm to compute \( P(q_i | E) \).

### Example

Given the variables Flu, Cold, Fever, Cough, and BodyAche, the structure of the Bayesian Network and the CPTs are defined. Inference is performed to compute the probability of having the Flu given that a patient has Fever and Cough.

---

This algorithmic description provides a structured approach to understanding and implementing Bayesian Networks, from defining the network structure and specifying conditional probabilities to performing inference and analyzing results.

# Bayesian Networks: A Tutorial

## Introduction

A Bayesian Network (BN) is a probabilistic graphical model that represents a set of variables and their conditional dependencies using a directed acyclic graph (DAG). Bayesian Networks are used for modeling uncertainty in complex systems and performing inference.

## Components of Bayesian Networks

1. **Nodes**: Represent random variables.
2. **Edges**: Represent conditional dependencies between variables.
3. **Conditional Probability Tables (CPTs)**: Define the probability of a node given its parents.

## Example Scenario

Consider a medical diagnosis problem where we want to model the relationship between diseases (Flu, Cold) and symptoms (Fever, Cough, BodyAche).

## Step-by-Step Implementation Using `pgmpy`

### Step 1: Install `pgmpy`

First, ensure you have `pgmpy` installed. You can install it using pip:

```bash
pip install pgmpy
```

### Step 2: Import Libraries

```python
import pandas as pd
import numpy as np
from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD
from pgmpy.inference import VariableElimination
```

### Step 3: Define the Bayesian Network Structure

Create a Bayesian network and define its structure (edges between nodes).

```python
# Define the structure
model = BayesianNetwork([('Flu', 'Fever'),
                         ('Cold', 'Fever'),
                         ('Flu', 'Cough'),
                         ('Cold', 'Cough'),
                         ('Flu', 'BodyAche'),
                         ('Cold', 'BodyAche')])
```

### Step 4: Define Conditional Probability Tables (CPTs)

Specify the CPTs for each node. This includes defining the probabilities of each node given its parents.

```python
# Define the CPD for Flu
cpd_flu = TabularCPD(variable='Flu', variable_card=2, values=[[0.9], [0.1]])

# Define the CPD for Cold
cpd_cold = TabularCPD(variable='Cold', variable_card=2, values=[[0.7], [0.3]])

# Define the CPD for Fever given Flu and Cold
cpd_fever = TabularCPD(variable='Fever', variable_card=2,
                       values=[[0.99, 0.9, 0.9, 0.0],
                               [0.01, 0.1, 0.1, 1.0]],
                       evidence=['Flu', 'Cold'], evidence_card=[2, 2])

# Define the CPD for Cough given Flu and Cold
cpd_cough = TabularCPD(variable='Cough', variable_card=2,
                       values=[[0.95, 0.3, 0.2, 0.0],
                               [0.05, 0.7, 0.8, 1.0]],
                       evidence=['Flu', 'Cold'], evidence_card=[2, 2])

# Define the CPD for BodyAche given Flu and Cold
cpd_body_ache = TabularCPD(variable='BodyAche', variable_card=2,
                           values=[[0.9, 0.5, 0.4, 0.1],
                                   [0.1, 0.5, 0.6, 0.9]],
                           evidence=['Flu', 'Cold'], evidence_card=[2, 2])

# Add CPDs to the model
model.add_cpds(cpd_flu, cpd_cold, cpd_fever, cpd_cough, cpd_body_ache)
```

### Step 5: Validate the Model

Check if the model is correctly defined.

```python
# Validate the model
assert model.check_model()
```

### Step 6: Perform Inference

Use the Variable Elimination method to perform inference on the Bayesian network.

```python
# Perform inference
inference = VariableElimination(model)

# Calculate the probability of having the flu given that the patient has a fever and cough
prob_flu_given_fever_cough = inference.query(variables=['Flu'], evidence={'Fever': 1, 'Cough': 1})
print(prob_flu_given_fever_cough)

# Calculate the probability of having a cold given that the patient has a fever and body ache
prob_cold_given_fever_body_ache = inference.query(variables=['Cold'], evidence={'Fever': 1, 'BodyAche': 1})
print(prob_cold_given_fever_body_ache)
```

### Step 7: Output the Inferred Probabilities

The results of the inference queries will provide the probabilities of the diseases given the symptoms.

## Conclusion

Bayesian Networks are powerful tools for modeling probabilistic relationships between variables and performing inference. By defining the structure, conditional probability tables, and using inference methods, we can effectively analyze complex systems with uncertainty.

This tutorial provides a basic overview of Bayesian Networks and their implementation using the `pgmpy` library in Python. You can expand this example by adding more variables, defining more complex dependencies, and performing more sophisticated inference tasks.

---

This detailed tutorial should help you understand and implement Bayesian Networks using the `pgmpy` library in Python. Feel free to customize and expand upon this example based on your specific use case and requirements.

In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pgmpy

%matplotlib notebook

For Task 7, you can use a standard dataset like the "Alarm" dataset, which is commonly used for Bayesian networks in medical diagnosis problems. However, creating a synthetic dataset that can be used to demonstrate the principles of Bayesian networks might be more educational and practical, especially for illustrating the implementation and inference process.

Here's a synthetic dataset for a medical diagnosis problem involving diseases and symptoms, suitable for building and testing a Bayesian network using the `pgmpy` library.

### Dataset Description

- **Nodes**: 
  - Diseases: Flu, Cold
  - Symptoms: Fever, Cough, Body Ache

- **Relationships**: 
  - Flu causes Fever and Body Ache
  - Cold causes Cough and sometimes Fever

### Sample Synthetic Data

| Flu | Cold | Fever | Cough | Body Ache |
|-----|------|-------|-------|-----------|
| 1   | 0    | 1     | 0     | 1         |
| 0   | 1    | 0     | 1     | 0         |
| 1   | 1    | 1     | 1     | 1         |
| 0   | 0    | 0     | 0     | 0         |
| ... | ...  | ...   | ...   | ...       |

### Steps to Create and Use the Dataset

1. **Create Synthetic Data**: Generate a dataset that simulates the relationships between diseases and symptoms.
2. **Define Bayesian Network Structure**: Specify the nodes and edges that represent the relationships.
3. **Specify Conditional Probability Tables (CPTs)**: Define the probabilities for each node given its parents.
4. **Perform Inference**: Calculate the probability of having a disease given some symptoms.

Here's how you can create and use this dataset:

### Explanation

1. **Generate Synthetic Data**: This script generates synthetic data for diseases (Flu, Cold) and symptoms (Fever, Cough, Body Ache) based on specified probabilities.
2. **Define Bayesian Network Structure**: It defines the structure of the Bayesian network representing the relationships between diseases and symptoms.
3. **Specify CPTs**: It sets the conditional probability tables for each node in the network.
4. **Perform Inference**: It uses the `pgmpy` library to perform inference, calculating the probability of having the flu given that a person has a fever and a cough.

You can use this synthetic dataset and the provided structure to implement and test a Bayesian network for the given task.

In [14]:
import pandas as pd
import numpy as np
from pgmpy.models import BayesianNetwork
from pgmpy.factors.discrete import TabularCPD
from pgmpy.inference import VariableElimination
from IPython.display import display, HTML

def create_synthetic_data(num_samples=1000, prob_flu=0.1, prob_cold=0.3):
    """
    Creates a synthetic dataset of diseases (Flu, Cold) and symptoms (Fever, Cough, Body Ache).
    
    Parameters:
    num_samples (int): Number of samples to generate.
    prob_flu (float): Probability of having the flu.
    prob_cold (float): Probability of having a cold.
    
    Returns:
    pd.DataFrame: DataFrame containing the generated data.
    """
    flu = np.random.binomial(1, prob_flu, num_samples)
    cold = np.random.binomial(1, prob_cold, num_samples)
    
    fever = np.random.binomial(1, np.clip(0.8*flu + 0.3*cold, 0, 1))
    cough = np.random.binomial(1, np.clip(0.2*flu + 0.7*cold, 0, 1))
    body_ache = np.random.binomial(1, np.clip(0.6*flu + 0.1*cold, 0, 1))
    
    data = pd.DataFrame({
        'Flu': flu,
        'Cold': cold,
        'Fever': fever,
        'Cough': cough,
        'BodyAche': body_ache
    })
    
    data.to_pickle('synthetic_disease_data.pkl')
    return data

def build_bayesian_network():
    """
    Builds a Bayesian Network model for diseases and symptoms.
    
    Returns:
    BayesianNetwork: The constructed Bayesian Network model.
    """
    model = BayesianNetwork([('Flu', 'Fever'), ('Flu', 'Body Ache'), ('Cold', 'Fever'), ('Cold', 'Cough')])

    cpd_flu = TabularCPD(variable='Flu', variable_card=2, values=[[0.8], [0.2]])
    cpd_cold = TabularCPD(variable='Cold', variable_card=2, values=[[0.7], [0.3]])

    cpd_fever = TabularCPD(variable='Fever', variable_card=2, 
                           values=[[0.9, 0.3, 0.1, 0.05],
                                   [0.1, 0.7, 0.9, 0.95]],
                           evidence=['Flu', 'Cold'],
                           evidence_card=[2, 2])

    cpd_cough = TabularCPD(variable='Cough', variable_card=2, 
                           values=[[0.9, 0.3], 
                                   [0.1, 0.7]],
                           evidence=['Cold'],
                           evidence_card=[2])

    cpd_body_ache = TabularCPD(variable='Body Ache', variable_card=2, 
                               values=[[0.7, 0.4], 
                                       [0.3, 0.6]],
                               evidence=['Flu'],
                               evidence_card=[2])

    model.add_cpds(cpd_flu, cpd_cold, cpd_fever, cpd_cough, cpd_body_ache)
    model.check_model()
    
    return model

def perform_inference(model, evidence):
    """
    Performs inference on the Bayesian Network model given some evidence.
    
    Parameters:
    model (BayesianNetwork): The Bayesian Network model.
    evidence (dict): Evidence for the inference.
    
    Returns:
    pd.Series: The result of the inference query as a pandas Series.
    """
    inference = VariableElimination(model)
    query_result = inference.query(variables=['Flu'], evidence=evidence)
    
    # Extract the result into a readable format
    result_dict = {f'P(Flu={state})': prob for state, prob in zip(query_result.state_names['Flu'], query_result.values)}
    result_series = pd.Series(result_dict)
    
    # Display evidence and inference result
    display(HTML("<h3>Evidence Provided</h3>"))
    display(pd.Series(evidence))
    display(HTML("<h3>Inference Result</h3>"))
    display(result_series)
    
    return result_series

def main():
    # Step 1: Create synthetic data
    data = create_synthetic_data()
    print('data head:')
    display(data.head())
    
    # Step 2: Build the Bayesian Network model
    model = build_bayesian_network()
    
    # Step 3: Perform inference
    evidence = {'Fever': 1, 'Cough': 1}
    query_result = perform_inference(model, evidence)

if __name__ == "__main__":
    main()


data head:


Flu  Cold  Fever  Cough  BodyAche
0    0     0      0      0         0
1    0     0      0      0         0
2    1     0      0      1         0
3    0     1      0      0         0
4    0     0      0      0         0

Fever    1
Cough    1
dtype: int64

P(Flu=0)    0.701195
P(Flu=1)    0.298805
dtype: float64

### High-Level Intuitive Interpretation

**Objective:**
We aimed to create a synthetic dataset to model the relationship between common illnesses (Flu and Cold) and their symptoms (Fever, Cough, and Body Ache). We then used this dataset to build a Bayesian Network to perform probabilistic inference, allowing us to determine the likelihood of having the flu given specific symptoms.

### Steps and Interpretations

1. **Data Generation:**
   - We generated a synthetic dataset with 1000 samples where each sample represents an individual.
   - For each individual, we determined whether they have the flu or a cold based on predefined probabilities (10% chance for flu, 30% chance for cold).
   - Based on whether an individual has the flu or a cold, we generated their symptoms (Fever, Cough, Body Ache) using probabilities that reflect how likely these symptoms are given the diseases.

   **Data Head:**
   ```
    Flu  Cold  Fever  Cough  BodyAche
    0    0     0      0      0
    0    0     0      0      0
    0    0     1      0      1
    0    0     0      0      0
    1    0     1      0      0
   ```

2. **Building the Bayesian Network:**
   - We constructed a Bayesian Network to model the dependencies between diseases and symptoms.
   - The network structure was defined such that:
     - Flu affects Fever and Body Ache.
     - Cold affects Fever and Cough.
   - Conditional Probability Tables (CPTs) were defined for each variable to quantify these dependencies.

3. **Performing Inference:**
   - We used the Variable Elimination algorithm to perform inference on the Bayesian Network.
   - Given the evidence that an individual has a fever (Fever=1) and a cough (Cough=1), we wanted to determine the probability of them having the flu.

   **Evidence Provided:**
   ```
   Fever    1
   Cough    1
   dtype: int64
   ```

   **Inference Result:**
   ```
   P(Flu=0)    0.701195
   P(Flu=1)    0.298805
   dtype: float64
   ```

### Intuitive Interpretation of Results

- **Evidence Provided:** We observed that the individual has both a fever and a cough.
- **Inference Result:**
  - The probability of the individual not having the flu (P(Flu=0)) is approximately 70.12%.
  - The probability of the individual having the flu (P(Flu=1)) is approximately 29.88%.

**Interpretation:**
Despite the presence of symptoms commonly associated with the flu (fever and cough), the model indicates that it is more likely (70.12%) that the individual does not have the flu. This result likely stems from the higher baseline probability of colds and their strong association with the given symptoms. The cold is a more common cause of these symptoms in our dataset, hence the higher probability of not having the flu.

### Conclusion

By using a Bayesian Network, we can integrate multiple pieces of evidence and quantify the likelihood of various hypotheses (in this case, having the flu) given the evidence. This approach helps in making informed decisions based on probabilistic reasoning, even when dealing with uncertain or incomplete information.

## Explanation of Variable Elimination

**Variable Elimination** is a fundamental algorithm used for performing inference in Bayesian Networks. The goal of variable elimination is to compute the probability distribution of a subset of variables (query variables) given some observed evidence while efficiently handling the summation over the other variables (hidden or irrelevant variables). Here's a high-level explanation of how variable elimination works:

### Key Concepts

1. **Bayesian Network:**
   - A graphical model that represents the probabilistic relationships among a set of variables.
   - Consists of nodes (variables) and directed edges (dependencies between variables).

2. **Inference:**
   - The process of computing the probability distribution of certain variables given evidence about other variables.

3. **Factors:**
   - Represent the joint probability distributions of subsets of variables.
   - In the context of Bayesian Networks, factors are derived from Conditional Probability Tables (CPTs).

### Steps of Variable Elimination

1. **Initialization:**
   - Identify the query variables $ \mathbf{Q} $ (variables of interest).
   - Identify the evidence variables $ \mathbf{E} $ (variables with known values).
   - Identify the hidden variables $ \mathbf{H} $ (all other variables that are not in $ \mathbf{Q} $ or $ \mathbf{E} $).

2. **Factors Representation:**
   - Convert the Bayesian Network's CPTs into factors.
   - Incorporate the evidence into the factors by setting the evidence variables to their observed values.

3. **Eliminate Hidden Variables:**
   - Select a hidden variable $ H $ and eliminate it by summing out $ H $ from the joint distribution.
   - This involves multiplying all factors that include $ H $ and then summing over $ H $.
   - Repeat this process for all hidden variables.

4. **Combine Remaining Factors:**
   - After all hidden variables are eliminated, combine the remaining factors that include the query variables.
   - Multiply these factors together to obtain the joint distribution over the query variables.

5. **Normalization:**
   - Normalize the resulting joint distribution to ensure it is a valid probability distribution.

### Example of Variable Elimination

Consider a simple Bayesian Network with three variables: $ A $, $ B $, and $ C $, with the following dependencies:
- $ A \rightarrow B $
- $ A \rightarrow C $

We want to perform inference to find $ P(A | B = b, C = c) $.

#### Steps:

1. **Initialization:**
   - Query variable: $ A $
   - Evidence variables: $ B, C $
   - Hidden variables: None in this case, as we are only interested in $ A $.

2. **Factors Representation:**
   - $ P(A) $ from the CPT of $ A $.
   - $ P(B | A) $ from the CPT of $ B $ given $ A $.
   - $ P(C | A) $ from the CPT of $ C $ given $ A $.

3. **Incorporate Evidence:**
   - Set $ B = b $ and $ C = c $ in the factors.

4. **Combine Factors:**
   - Multiply the factors $ P(A) $, $ P(B = b | A) $, and $ P(C = c | A) $ to obtain the joint distribution over $ A $.

5. **Normalization:**
   - Normalize the resulting distribution to obtain $ P(A | B = b, C = c) $.


### Summary

**Variable Elimination** is an efficient method for performing inference in Bayesian Networks. It systematically eliminates hidden variables by summing them out, combines remaining factors, and normalizes the result to provide the probability distribution of the query variables given the evidence. This approach allows for accurate probabilistic reasoning even in complex networks with many variables.